File: crates/gen-guest-rust/tests/flegs.rs
pub mod flegs {
    use ::tauri_bindgen_guest_rust::bitflags;
    use ::tauri_bindgen_guest_rust::tauri_bindgen_abi;
    bitflags::bitflags! {
        #[derive(tauri_bindgen_abi::Writable, tauri_bindgen_abi::Readable)] pub struct
        Flag1 : u8 { const B0 = 1 << 0; }
    }
    bitflags::bitflags! {
        #[derive(tauri_bindgen_abi::Writable, tauri_bindgen_abi::Readable)] pub struct
        Flag16 : u16 { const B0 = 1 << 0; const B1 = 1 << 1; const B2 = 1 << 2; const B3
        = 1 << 3; const B4 = 1 << 4; const B5 = 1 << 5; const B6 = 1 << 6; const B7 = 1
        << 7; const B8 = 1 << 8; const B9 = 1 << 9; const B10 = 1 << 10; const B11 = 1 <<
        11; const B12 = 1 << 12; const B13 = 1 << 13; const B14 = 1 << 14; const B15 = 1
        << 15; }
    }
    bitflags::bitflags! {
        #[derive(tauri_bindgen_abi::Writable, tauri_bindgen_abi::Readable)] pub struct
        Flag2 : u8 { const B0 = 1 << 0; const B1 = 1 << 1; }
    }
    bitflags::bitflags! {
        #[derive(tauri_bindgen_abi::Writable, tauri_bindgen_abi::Readable)] pub struct
        Flag32 : u32 { const B0 = 1 << 0; const B1 = 1 << 1; const B2 = 1 << 2; const B3
        = 1 << 3; const B4 = 1 << 4; const B5 = 1 << 5; const B6 = 1 << 6; const B7 = 1
        << 7; const B8 = 1 << 8; const B9 = 1 << 9; const B10 = 1 << 10; const B11 = 1 <<
        11; const B12 = 1 << 12; const B13 = 1 << 13; const B14 = 1 << 14; const B15 = 1
        << 15; const B16 = 1 << 16; const B17 = 1 << 17; const B18 = 1 << 18; const B19 =
        1 << 19; const B20 = 1 << 20; const B21 = 1 << 21; const B22 = 1 << 22; const B23
        = 1 << 23; const B24 = 1 << 24; const B25 = 1 << 25; const B26 = 1 << 26; const
        B27 = 1 << 27; const B28 = 1 << 28; const B29 = 1 << 29; const B30 = 1 << 30;
        const B31 = 1 << 31; }
    }
    bitflags::bitflags! {
        #[derive(tauri_bindgen_abi::Writable, tauri_bindgen_abi::Readable)] pub struct
        Flag4 : u8 { const B0 = 1 << 0; const B1 = 1 << 1; const B2 = 1 << 2; const B3 =
        1 << 3; }
    }
    bitflags::bitflags! {
        #[derive(tauri_bindgen_abi::Writable, tauri_bindgen_abi::Readable)] pub struct
        Flag64 : u64 { const B0 = 1 << 0; const B1 = 1 << 1; const B2 = 1 << 2; const B3
        = 1 << 3; const B4 = 1 << 4; const B5 = 1 << 5; const B6 = 1 << 6; const B7 = 1
        << 7; const B8 = 1 << 8; const B9 = 1 << 9; const B10 = 1 << 10; const B11 = 1 <<
        11; const B12 = 1 << 12; const B13 = 1 << 13; const B14 = 1 << 14; const B15 = 1
        << 15; const B16 = 1 << 16; const B17 = 1 << 17; const B18 = 1 << 18; const B19 =
        1 << 19; const B20 = 1 << 20; const B21 = 1 << 21; const B22 = 1 << 22; const B23
        = 1 << 23; const B24 = 1 << 24; const B25 = 1 << 25; const B26 = 1 << 26; const
        B27 = 1 << 27; const B28 = 1 << 28; const B29 = 1 << 29; const B30 = 1 << 30;
        const B31 = 1 << 31; const B32 = 1 << 32; const B33 = 1 << 33; const B34 = 1 <<
        34; const B35 = 1 << 35; const B36 = 1 << 36; const B37 = 1 << 37; const B38 = 1
        << 38; const B39 = 1 << 39; const B40 = 1 << 40; const B41 = 1 << 41; const B42 =
        1 << 42; const B43 = 1 << 43; const B44 = 1 << 44; const B45 = 1 << 45; const B46
        = 1 << 46; const B47 = 1 << 47; const B48 = 1 << 48; const B49 = 1 << 49; const
        B50 = 1 << 50; const B51 = 1 << 51; const B52 = 1 << 52; const B53 = 1 << 53;
        const B54 = 1 << 54; const B55 = 1 << 55; const B56 = 1 << 56; const B57 = 1 <<
        57; const B58 = 1 << 58; const B59 = 1 << 59; const B60 = 1 << 60; const B61 = 1
        << 61; const B62 = 1 << 62; const B63 = 1 << 63; }
    }
    bitflags::bitflags! {
        #[derive(tauri_bindgen_abi::Writable, tauri_bindgen_abi::Readable)] pub struct
        Flag8 : u8 { const B0 = 1 << 0; const B1 = 1 << 1; const B2 = 1 << 2; const B3 =
        1 << 3; const B4 = 1 << 4; const B5 = 1 << 5; const B6 = 1 << 6; const B7 = 1 <<
        7; }
    }
    pub async fn roundtrip_flag1(x: Flag1) -> Flag1 {
        todo!()
    }
    pub async fn roundtrip_flag2(x: Flag2) -> Flag2 {
        todo!()
    }
    pub async fn roundtrip_flag4(x: Flag4) -> Flag4 {
        todo!()
    }
    pub async fn roundtrip_flag8(x: Flag8) -> Flag8 {
        todo!()
    }
    pub async fn roundtrip_flag16(x: Flag16) -> Flag16 {
        todo!()
    }
    pub async fn roundtrip_flag32(x: Flag32) -> Flag32 {
        todo!()
    }
    pub async fn roundtrip_flag64(x: Flag64) -> Flag64 {
        todo!()
    }
}

File: crates/gen-guest-ts/tests/flegs.ts
class Deserializer {
    source
    offset
    
    constructor(bytes) {
        this.source = bytes
        this.offset = 0
    }

    pop() {
        return this.source[this.offset++]
    }

    try_take_n(len) {
        const out = this.source.slice(this.offset, this.offset + len)
        this.offset += len
        return out
    }
}
function varint_max(type) {
    const BITS_PER_BYTE = 8;
    const BITS_PER_VARINT_BYTE = 7;

    const bits = type * BITS_PER_BYTE;

    const roundup_bits = bits + (BITS_PER_BYTE - 1);

    return Math.floor(roundup_bits / BITS_PER_VARINT_BYTE);
}
function max_of_last_byte(type) {
  let extra_bits = type % 7;
  return (1 << extra_bits) - 1;
}

function de_varint(de, type) {
  let out = 0;

  for (let i = 0; i < varint_max(type); i++) {
    const val = de.pop();
    const carry = val & 0x7F;
    out |= carry << (7 * i);

    if ((val & 0x80) === 0) {
      if (i === varint_max(type) - 1 && val > max_of_last_byte(type)) {
        throw new Error('deserialize bad variant')
      } else {
        return out
      }
    }
  }

  throw new Error('deserialize bad variant')
}

function de_varint_big(de, type) {
  let out = 0n;

  for (let i = 0; i < varint_max(type); i++) {
    const val = de.pop();
    const carry = BigInt(val) & 0x7Fn;
    out |= carry << (7n * BigInt(i));

    if ((val & 0x80) === 0) {
      if (i === varint_max(type) - 1 && val > max_of_last_byte(type)) {
        throw new Error('deserialize bad variant')
      } else {
        return out
      }
    }
  }

  throw new Error('deserialize bad variant')
}
function deserializeU8(de) {
    return de.pop()
}
function deserializeU16(de) {
    return de_varint(de, 16)
}
function deserializeU32(de) {
    return de_varint(de, 32)
}
function deserializeU64(de) {
  return de_varint_big(de, 64)
}
function ser_varint(out, type, val) {
  let buf = []
  for (let i = 0; i < varint_max(type); i++) {
    const buffer = new ArrayBuffer(type / 8);
    const view = new DataView(buffer);
    view.setInt16(0, val, true);
    buf[i] = view.getUint8(0);
    if (val < 128) {
      out.push(...buf)
      return;
    }

    buf[i] |= 0x80;
    val >>= 7;
  }
  out.push(...buf)
}

function ser_varint_big(out, type, val) {
  let buf = []
  for (let i = 0; i < varint_max(type); i++) {
    const buffer = new ArrayBuffer(type / 8);
    const view = new DataView(buffer);
    view.setInt16(0, Number(val), true);
    buf[i] = view.getUint8(0);
    if (val < 128) {
      out.push(...buf)
      return;
    }

    buf[i] |= 0x80;
    val >>= 7n;
  }
  out.push(...buf)
}
function serializeU8(out, val) {
    return out.push(val)
}
function serializeU16(out, val) {
    return ser_varint(out, 16, val)
}
function serializeU32(out, val) {
    return ser_varint(out, 32, val)
}
function serializeU64(out, val) {
  return ser_varint_big(out, 64, BigInt(val))
}
function deserializeFlag1(de) {
    return deserializeU8(de)
}function deserializeFlag2(de) {
    return deserializeU8(de)
}function deserializeFlag4(de) {
    return deserializeU8(de)
}function deserializeFlag8(de) {
    return deserializeU8(de)
}function deserializeFlag16(de) {
    return deserializeU16(de)
}function deserializeFlag32(de) {
    return deserializeU32(de)
}function deserializeFlag64(de) {
    return deserializeU64(de)
}function serializeFlag1(out, val) {
    return serializeU8(out, val)
}function serializeFlag2(out, val) {
    return serializeU8(out, val)
}function serializeFlag4(out, val) {
    return serializeU8(out, val)
}function serializeFlag8(out, val) {
    return serializeU8(out, val)
}function serializeFlag16(out, val) {
    return serializeU16(out, val)
}function serializeFlag32(out, val) {
    return serializeU32(out, val)
}function serializeFlag64(out, val) {
    return serializeU64(out, val)
}

export enum Flag1 { 
B0 = 2,
 }

export enum Flag2 { 
B0 = 2,

B1 = 4,
 }

export enum Flag4 { 
B0 = 2,

B1 = 4,

B2 = 8,

B3 = 16,
 }

export enum Flag8 { 
B0 = 2,

B1 = 4,

B2 = 8,

B3 = 16,

B4 = 32,

B5 = 64,

B6 = 128,

B7 = 256,
 }

export enum Flag16 { 
B0 = 2,

B1 = 4,

B2 = 8,

B3 = 16,

B4 = 32,

B5 = 64,

B6 = 128,

B7 = 256,

B8 = 512,

B9 = 1024,

B10 = 2048,

B11 = 4096,

B12 = 8192,

B13 = 16384,

B14 = 32768,

B15 = 65536,
 }

export enum Flag32 { 
B0 = 2,

B1 = 4,

B2 = 8,

B3 = 16,

B4 = 32,

B5 = 64,

B6 = 128,

B7 = 256,

B8 = 512,

B9 = 1024,

B10 = 2048,

B11 = 4096,

B12 = 8192,

B13 = 16384,

B14 = 32768,

B15 = 65536,

B16 = 131072,

B17 = 262144,

B18 = 524288,

B19 = 1048576,

B20 = 2097152,

B21 = 4194304,

B22 = 8388608,

B23 = 16777216,

B24 = 33554432,

B25 = 67108864,

B26 = 134217728,

B27 = 268435456,

B28 = 536870912,

B29 = 1073741824,

B30 = 2147483648,

B31 = 4294967296,
 }

export enum Flag64 { 
B0 = 2,

B1 = 4,

B2 = 8,

B3 = 16,

B4 = 32,

B5 = 64,

B6 = 128,

B7 = 256,

B8 = 512,

B9 = 1024,

B10 = 2048,

B11 = 4096,

B12 = 8192,

B13 = 16384,

B14 = 32768,

B15 = 65536,

B16 = 131072,

B17 = 262144,

B18 = 524288,

B19 = 1048576,

B20 = 2097152,

B21 = 4194304,

B22 = 8388608,

B23 = 16777216,

B24 = 33554432,

B25 = 67108864,

B26 = 134217728,

B27 = 268435456,

B28 = 536870912,

B29 = 1073741824,

B30 = 2147483648,

B31 = 4294967296,

B32 = 8589934592,

B33 = 17179869184,

B34 = 34359738368,

B35 = 68719476736,

B36 = 137438953472,

B37 = 274877906944,

B38 = 549755813888,

B39 = 1099511627776,

B40 = 2199023255552,

B41 = 4398046511104,

B42 = 8796093022208,

B43 = 17592186044416,

B44 = 35184372088832,

B45 = 70368744177664,

B46 = 140737488355328,

B47 = 281474976710656,

B48 = 562949953421312,

B49 = 1125899906842624,

B50 = 2251799813685248,

B51 = 4503599627370496,

B52 = 9007199254740992,

B53 = 18014398509481984,

B54 = 36028797018963968,

B55 = 72057594037927936,

B56 = 144115188075855872,

B57 = 288230376151711744,

B58 = 576460752303423488,

B59 = 1152921504606846976,

B60 = 2305843009213693952,

B61 = 4611686018427387904,

B62 = 9223372036854775808,

B63 = 0,
 }



export async function roundtripFlag1 (x: Flag1) : Promise<Flag1> {
    const out = []
    serializeFlag1(out, x)
    
    return fetch('ipc://localhost/flegs/roundtrip_flag1', { method: "POST", body: Uint8Array.from(out) })
        .then(r => r.arrayBuffer())
        .then(bytes => {
            const de = new Deserializer(new Uint8Array(bytes))

            return deserializeFlag1(de)
        }) as Promise<Flag1>
}
        

export async function roundtripFlag2 (x: Flag2) : Promise<Flag2> {
    const out = []
    serializeFlag2(out, x)
    
    return fetch('ipc://localhost/flegs/roundtrip_flag2', { method: "POST", body: Uint8Array.from(out) })
        .then(r => r.arrayBuffer())
        .then(bytes => {
            const de = new Deserializer(new Uint8Array(bytes))

            return deserializeFlag2(de)
        }) as Promise<Flag2>
}
        

export async function roundtripFlag4 (x: Flag4) : Promise<Flag4> {
    const out = []
    serializeFlag4(out, x)
    
    return fetch('ipc://localhost/flegs/roundtrip_flag4', { method: "POST", body: Uint8Array.from(out) })
        .then(r => r.arrayBuffer())
        .then(bytes => {
            const de = new Deserializer(new Uint8Array(bytes))

            return deserializeFlag4(de)
        }) as Promise<Flag4>
}
        

export async function roundtripFlag8 (x: Flag8) : Promise<Flag8> {
    const out = []
    serializeFlag8(out, x)
    
    return fetch('ipc://localhost/flegs/roundtrip_flag8', { method: "POST", body: Uint8Array.from(out) })
        .then(r => r.arrayBuffer())
        .then(bytes => {
            const de = new Deserializer(new Uint8Array(bytes))

            return deserializeFlag8(de)
        }) as Promise<Flag8>
}
        

export async function roundtripFlag16 (x: Flag16) : Promise<Flag16> {
    const out = []
    serializeFlag16(out, x)
    
    return fetch('ipc://localhost/flegs/roundtrip_flag16', { method: "POST", body: Uint8Array.from(out) })
        .then(r => r.arrayBuffer())
        .then(bytes => {
            const de = new Deserializer(new Uint8Array(bytes))

            return deserializeFlag16(de)
        }) as Promise<Flag16>
}
        

export async function roundtripFlag32 (x: Flag32) : Promise<Flag32> {
    const out = []
    serializeFlag32(out, x)
    
    return fetch('ipc://localhost/flegs/roundtrip_flag32', { method: "POST", body: Uint8Array.from(out) })
        .then(r => r.arrayBuffer())
        .then(bytes => {
            const de = new Deserializer(new Uint8Array(bytes))

            return deserializeFlag32(de)
        }) as Promise<Flag32>
}
        

export async function roundtripFlag64 (x: Flag64) : Promise<Flag64> {
    const out = []
    serializeFlag64(out, x)
    
    return fetch('ipc://localhost/flegs/roundtrip_flag64', { method: "POST", body: Uint8Array.from(out) })
        .then(r => r.arrayBuffer())
        .then(bytes => {
            const de = new Deserializer(new Uint8Array(bytes))

            return deserializeFlag64(de)
        }) as Promise<Flag64>
}
        